server/repository/src/db_diesel/changelog/test.rs

```diff
@@ -0,0 +1,572 @@
+use super::changelog::changelog::dsl as changelog_dsl;
+use diesel::prelude::*;
+use util::{inline_edit, inline_init};
+
+use crate::{
+    mock::{
+        mock_item_a, mock_name_a, mock_name_b, mock_name_store_a, mock_name_store_b, MockData,
+        MockDataInserts,
+    },
+    test_db::{self, setup_all, setup_all_with_data},
+    ChangelogAction, ChangelogFilter, ChangelogRepository, ChangelogRow, ChangelogTableName,
+    EqualFilter, InvoiceLineRow, InvoiceLineRowRepository, InvoiceRow, InvoiceRowRepository,
+    NameRow, NameRowRepository, RequisitionLineRow, RequisitionLineRowRepository, RequisitionRow,
+    RequisitionRowRepository, StorageConnection, StoreRow,
+};
+
+#[actix_rt::test]
+async fn test_changelog() {
+    let (_, connection, _, _) = test_db::setup_all("test_changelog", MockDataInserts::none()).await;
+
+    // use name entries to populate the changelog (via the trigger)
+    let name_repo = NameRowRepository::new(&connection);
+    let repo = ChangelogRepository::new(&connection);
+
+    // single entry:
+    let name_a = mock_name_a();
+    name_repo.upsert_one(&name_a).unwrap();
+    let mut result = repo.changelogs(0, 10, None).unwrap();
+    assert_eq!(1, result.len());
+    let log_entry = result.pop().unwrap();
+    assert_eq!(
+        log_entry,
+        inline_init(|r: &mut ChangelogRow| {
+            r.cursor = 1;
+            r.table_name = ChangelogTableName::Name;
+            r.record_id = name_a.id.clone();
+            r.row_action = ChangelogAction::Upsert;
+        })
+    );
+
+    // querying from the first entry should give the same result:
+    assert_eq!(
+        repo.changelogs(0, 10, None).unwrap(),
+        repo.changelogs(1, 10, None).unwrap()
+    );
+
+    // update the entry
+    let mut name_a_update = mock_name_a();
+    name_a_update.comment = Some("updated".to_string());
+    name_repo.upsert_one(&name_a_update).unwrap();
+    let mut result = repo
+        .changelogs((log_entry.cursor + 1) as u64, 10, None)
+        .unwrap();
+    assert_eq!(1, result.len());
+    let log_entry = result.pop().unwrap();
+    assert_eq!(
+        log_entry,
+        inline_init(|r: &mut ChangelogRow| {
+            r.cursor = 2;
+            r.table_name = ChangelogTableName::Name;
+            r.record_id = name_a.id.clone();
+            r.row_action = ChangelogAction::Upsert;
+        })
+    );
+
+    // query the full list from cursor=0
```

Comment:
```suggestion
    // query the full list from cursor=0
    // because we use the changelog_deduped view, we should only get the latest changelog row for the record_id
```

Reply:
thanks (I didn't end up reviewing those test, just relocated)
oops, I committed rather then using suggestion: 43dafb160173ce6749dfcf449929245692c682a5